KAFKA-10357: Refactor makeReady to reduce complexity and support upcoming KIP-698 changes

k-apol opened the pull request:
This PR reduces the cyclomatic complexity of the makeReady method by
extracting parts of its logic into smaller, well-named helper methods.
This structural cleanup makes the code easier to understand and
maintain.

These changes are intended to prepare for upcoming work related to
[KIP-698](https://cwiki.apache.org/confluence/display/KAFKA/KIP-698),
which will introduce additional branching logic to makeReady.
Refactoring now ensures that future modifications can be made in a more
modular and testable way.

Reviewer: Matthias J. Sax <matthias@confluent.io>

k-apol committed:
Clean rebase from previous PR

k-apol committed:
refactor naming convention to be more clear - we are computing topics to create, so we should return those

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java:
```diff
@@ -113,7 +113,7 @@ public InternalTopicManager(final Time time,
     }
 
-    static class ValidationResult {
+    public static class ValidationResult {
```
Why do we need to make this class `public` ?

k-apol replied:
I will need to use it from inside `KafkaStreams.java` - the new `init()` method will need to make decisions based off of the validation results of the passed topic config, and I made it public here so that I will be able to do so on the next PR. I suppose it can wait until the actual implementation of `init()` if you would prefer, or maybe you have an alternative approach? 

[Here is the code I am referencing from the (now closed) previous PR](https://github.com/k-apol/kafka/pull/1/files#diff-ab27af136b0c45ed402ec44368a91380b018bf06f1a9722324fe6be8d5220f7dR388)

mjsax replied:
Thanks. Ok to leave as-is as we know the context. In general, it would be cleaner to make it public on the other PR as the other PR contains the context why we make it public. But it's ok as-is, too.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java:
```diff
@@ -461,120 +466,119 @@ public Set<String> makeReady(final Map<String, InternalTopicConfig> topics) {
+    }
+
+    private Set<NewTopic> computeTopicsToCreate(final Map<String, InternalTopicConfig> topics,
+                                                      final Set<String> topicsNotReady,
```
nit: fix indention

k-apol replied:
Thank you, will push an update for this

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java:
```diff
@@ -461,120 +466,119 @@ public Set<String> makeReady(final Map<String, InternalTopicConfig> topics) {
-                    retryBackOffMs,
-                    deadlineMs - currentWallClockMs
-                );
-                Utils.sleep(retryBackOffMs);
```
It seems this retryBackOffMs got lost? (Including the INFO log?)

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java:
```diff
@@ -461,120 +466,119 @@ public Set<String> makeReady(final Map<String, InternalTopicConfig> topics) {
+                    String.format("Could not create topics within %d milliseconds. This can happen if the Kafka cluster is temporarily not available.", retryTimeoutMs),
+                    null
+                ));
 
```
Cf my other comment further below: I think we need to add the retryBackOff sleep here, if we where not able to create all topics yet, but did not yet hit the timeout.

k-apol replied:
Pushing an update for this, adding the log line and backoff back (I did not mean to remove this, thanks for catching it)

I want to exponentially backoff here so that I am not spamming the broker with requests to create the topics right? If we didn't have this here, we could potentially overload our broker?

k-apol replied:
Added this one back, resolving

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java:
```diff
@@ -461,120 +466,119 @@ public Set<String> makeReady(final Map<String, InternalTopicConfig> topics) {
+                                                      final Set<String> topicsNotReady,
+                                                      final Set<String> newTopics) {
+        final Set<String> tempUnknownTopics = new HashSet<>();
+        final Set<String> validatedTopics = validateTopics(topicsNotReady, topics, tempUnknownTopics);
```
Doesn't `validateTopics` return a set of topic it could _not_ validate, and still need to get created?

k-apol replied:
Yes, you're right.. maybe we call it `validateTopicResult ` instead?

k-apol replied:
The return value represents topics that are configured correctly, and that also need to get created

The method call here iterates the provided topics, and validates three things:

1. The `topicsMap `provided in the config knows about all of the topics we are trying to validate
2. The number of partitions for each internal topic is defined
3. If the internal topic already exists, we decide **not** to add it to the return set but validate the number of partitions it has matches what the `InternalTopicManager` has

If all 1 and 2 pass, we add to the return set, and repeat for all topics passed into the function

mjsax replied:
> The return value represents topics that are configured correctly, and that also need to get created

Well, the method does validate topics **_if_** they exist. That's why the method is called `validateTopics` -- but topics that do not exist, can by definition not be validated, but need to be created. So the returned set of topics, as you also confirmed (`also need to get created`), are non-existing topics, and thus, they did not get validated as there is nothing to be done ("validation" mean, checks if an existing topic meets the expected partition count).

Does this make sense?

k-apol replied:
Yes.. I see your point, it may be a good opportunity to rename. If it's not properly validating the topics, I don't think we should name is `validateTopics`, this is throwing me off. 

The purpose of this method is to verify the partition count of the existing topics and return topics that have not yet been created.. maybe `findTopicsToCreate`, `determineTopicsToCreate`, `computeMissingTopics`.. do any of these sound good to you?

mjsax replied:
Well, the method does two things, validate existing topics and identify non-existing ones. So I think both names are ok, the current `validateTopics` or any of the names you proposed. As the method does two thing, there will be some side-effect. I am fine either way.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java:
```diff
@@ -461,120 +466,119 @@ public Set<String> makeReady(final Map<String, InternalTopicConfig> topics) {
-        Set<String> topicsNotReady = new HashSet<>(topics.keySet());
-        final Set<String> newlyCreatedTopics = new HashSet<>();
+        final Set<String> topicsNotReady = new HashSet<>(topics.keySet());
+        final Set<String> newTopics = new HashSet<>();
```
Why do you change the name? To me, `newlyCreatedTopics` sounds like a very good name?

k-apol replied:
My reasoning was that they aren't created until we call the admin client, at this point in the process they just describe the topics we want to create.

mjsax replied:
But `newTopics` is created empty here, and below we use `topicsToCreate` for the topics we want to create? So I am confused now? Why would we track the same thing twice. -> And why do we pass in` newTopics` into `computeTopicsToCreate` _and_ return a set of `topicToCreate` ?

mjsax replied:
Re-reading the code again, it seems the control flow is still to convoluted?

We have method `createTopics(...)` in which we pass `topicsToCreate` (this makes sense), but we also pass in `topicsNotReady` and modify it as a side effect -- this does not seem to be clean? It might be better to let `createTopics(...)` return the set of topic it could create, and do the overall bookkeeping outside of `createTopics(...)` (we could not need to pass `topicsNotReady` into `createTopics(...)` at all for this case, but do:
```
final Set<String> createdTopics = createTopics(topicsToCreate, deadlineMs);
topicsNotReady.removeAll(createdTopics);
```

mjsax replied:
Next, we have `computeTopicsToCreate(...)` -- here we pass in `newTopics` which in the end tracks (over all iterations we need), all topics we do create -- the difference to the return value of `computeTopicsToCreate(...)` is that the method may return a subset of topic only, which we would create in the current iteration.

Again, it seems not to be clean to pass `newTopics` into `computeTopicsToCreate(...)` to begin with, and we should instead add newly created topic after they where created, so maybe we should do:
```
final Set<String> createdTopics = createTopics(topicsToCreate, deadlineMs); topicsNotReady.removeAll(createdTopics);
newTopic.addAll(createdTopics); // for this case, it would make sense not to rename the variable to `newlyCreatedTopics` as I proposed originally?
```

mjsax replied:
In the end, both `createTopics(...)` and `computeTopicsToCreate(...)` imply side effects, for book keeping, but these side effects make it hard to actually reason about the control flow...

Thoughts?

k-apol replied:
> Re-reading the code again, it seems the control flow is still to convoluted?
> 
> We have method `createTopics(...)` in which we pass `topicsToCreate` (this makes sense), but we also pass in `topicsNotReady` and modify it as a side effect -- this does not seem to be clean? It might be better to let `createTopics(...)` return the set of topic it could create, and do the overall bookkeeping outside of `createTopics(...)` (we could not need to pass `topicsNotReady` into `createTopics(...)` at all for this case, but do:
> 
> ```
> final Set<String> createdTopics = createTopics(topicsToCreate, deadlineMs);
> topicsNotReady.removeAll(createdTopics);
> ```

I agree, this is hard to reason about. I propose we pass `topicsNotReady` into `createTopics(topicsToCreate, topicsNotReady, deadlineMs)`, and just abstract the bookeeping step away inside the method. It feels okay to modify the `topicsNotReady` inside here, since it logically tracks that as we create a topic, we have one less that is 'not ready'.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java:
```diff
@@ -461,120 +466,119 @@ public Set<String> makeReady(final Map<String, InternalTopicConfig> topics) {
-
+            final Set<NewTopic> topicsToCreate = computeTopicsToCreate(topics, topicsNotReady, newTopics);
+            if (!topicsToCreate.isEmpty()) {
+                readyTopics(topicsToCreate, topicsNotReady);
```
Maybe rename `readyTopics(...)` to `createTopics(...)` ?

k-apol replied:
The admin client call we do to create them has the same name, I agree it better represents what is happening but I felt it might be confusing to use the same method name, even if it's in a different class. 

`readyTopics(...)` convention ties it a bit to `makeReady(...)`, it seemed like the next best alternative. If you feel that the 'duplicate' naming with `createTopics` is fine, then I will update it to this

k-apol replied:
Looking at this again, It's fine to do the naming like this, just a convention I do not see as often with my daily-use language of Typescript so it made me scratch my head a bit. Pushing an update for this and resolving

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java:
```diff
@@ -461,120 +466,119 @@ public Set<String> makeReady(final Map<String, InternalTopicConfig> topics) {
+        final Set<String> tempUnknownTopics = new HashSet<>();
+        final Set<String> validatedTopics = validateTopics(topicsNotReady, topics, tempUnknownTopics);
+
+        newTopics.addAll(validatedTopics);
```
Cf comment above: I think this line would currently add topic with are not read to `newTopics` what seems incorrect?

k-apol replied:
Added some thoughts

k-apol committed:
Rename readyTopics -> createTopics
Add log line for incomplete topic creation call to broker
Pass deadline parameter into createTopics for proper timeout handling
Add exponential backoff back to createTopics (removed by mistake)

k-apol committed:
Refactor: Simplify control flow and update naming conventions

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java:
```diff
@@ -527,8 +527,9 @@ private Set<NewTopic> computeTopicsToCreate(final Map<String, InternalTopicConfi
-    private void readyTopics(final Set<NewTopic> topicsToCreate,
-                             final Set<String> topicsNotReady) {
+    private void createTopics(final Set<NewTopic> topicsToCreate,
+                             final Set<String> topicsNotReady,
```
nit: indention

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java:
```diff
@@ -578,8 +579,29 @@ private void readyTopics(final Set<NewTopic> topicsToCreate,
                 }
             }
+
+            if (!topicsNotReady.isEmpty()) {
```
Should this go after the for-loop? I don't think we need to retry-backoff after verifying a single topic, but should be sufficient to check once for the whole "batch" of topics?

mjsax commented:
It seems `InternalTopicManagerTest` fails. This refactoring should not change the logic though. Can you take a look? Can make another full pass, after the test is back to green :)

k-apol committed:
Update error messages to new timeout context format

k-apol committed:
Add logic for when leader is unavailable to indicate whether a topic exists. We can cleanly determine no work is left based on this and topicsToCreate result, allowing us to stop looping early

k-apol committed:
Update test messages to new timeoutException format - this accommodates the common method created to throw a timeout, but gives context to what kind of timeout happened. Looking for feedback on whether this is correct level of detail for given context

k-apol committed:
cleanup

k-apol committed:
Move second stub to avoid repeated "failure" response (was causing repeated looping)

k-apol committed:
Remove misleading code - in this case, we did not "newly create" the topics, though we can stop iteration.

mjsax commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/InternalTopicManagerTest.java:
```diff
@@ -782,16 +782,18 @@ public void shouldCompleteTopicValidationOnRetry() {
+                        mkEntry(topic2, topicDescriptionFailFuture) // first call: missing
+                )))
+                .thenAnswer(answer -> new MockDescribeTopicsResult(mkMap(
+                        mkEntry(topic1, topicDescriptionSuccessFuture),
```
Why do we return topic1 here again? The original code only returned `topic2` in the result.

k-apol replied:
So when tracing this test to see why it was failing, it was retrying topic 2 repeatedly (which is always throwing the `TopicExistsException`, as the stub indicates), but it was never exiting that loop. When mocking the time so that it will deadline, I ultimately got a timeout exception (which I think is fine here?).

I felt it was better to mock it succeeding the second call than changing the expected test output by expecting the timeout exception. It still models a transient error, and is still forced to exercise the retry logic that I feel this test is looking to validate.

mjsax replied:
I was looking into this with a debugger, and it seems the current rewrite introduces a small regression (making this change necessary). In the old code, for both topics we query the number of partitions, and get an error for topic_2 -- but because we can get information for topic_1, we make it "as ready" (cf my other comment). In the new code, we "forget" to mark `topic_1` and ready, and thus, on retry for `topic_2`, we query the number of partitions for both topics again (while it would not be necessary to query `topic_1` a second time).

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java:
```diff
@@ -461,121 +466,155 @@ public Set<String> makeReady(final Map<String, InternalTopicConfig> topics) {
 
         while (!topicsNotReady.isEmpty()) {
             final Set<String> tempUnknownTopics = new HashSet<>();
-            topicsNotReady = validateTopics(topicsNotReady, topics, tempUnknownTopics);
```
Cf my other commend on the test. Here, he old code update `topicNotReady` and the new map only contains `topic_2` after the first validation try. -- Below in the new code, we do not make this update though.

mjsax commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/InternalTopicManagerTest.java:
```diff
@@ -782,16 +782,18 @@ public void shouldCompleteTopicValidationOnRetry() {
                                  mkEntry(TopicConfig.MESSAGE_TIMESTAMP_TYPE_CONFIG, "CreateTime"))))))
             .thenAnswer(answer -> new MockCreateTopicsResult(Collections.singletonMap(topic2, topicCreationFuture)));
-        when(admin.describeTopics(Collections.singleton(topic2)))
-            .thenAnswer(answer -> new MockDescribeTopicsResult(Collections.singletonMap(topic2, topicDescriptionSuccessFuture)));
```
We should remove the code change from above, and keep this one, because on retry, we should only query for `topic2`.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java:
```diff
@@ -461,121 +466,155 @@ public Set<String> makeReady(final Map<String, InternalTopicConfig> topics) {
+            final Map<String, InternalTopicConfig> notReadyTopicsMap = topics.entrySet().stream()
+                    .filter(e -> topicsNotReady.contains(e.getKey()))
+                    .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+            final Set<NewTopic> topicsToCreate = computeTopicsToCreate(notReadyTopicsMap, tempUnknownTopics);
```
Here we compute `topicToCreate`, but for all existing topics, we don't return them, but we also don't remove existing topics from `topicsNotReady`.

I think, we need to add a step `topicsNotReady.retainAll(notReadyTopicsMap.keySet())` after this line, plus modify `notReadyTopicsMap` inside `computeTopicsToCreate` and remove all entries for the topics we successfully verify.

k-apol replied:
done in recent commit, resolving

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java:
```diff
@@ -673,11 +708,11 @@ private Set<String> validateTopics(final Set<String> topicsToValidate,
                     throw new StreamsException(errorMsg);
                 }
```
If the `if` above does not pass, ie, the number of partitions is as expected and we don't throw an error, we can remove the move the current topic from `topicsMap`, as we successfully remove it. This is the side effect which avoid that we query for the same topic multiple times.

k-apol replied:
done in recent commit, this was the key to solving the regression that I introduced. resolving this

k-apol committed:
Retain uknown and uncreated topics while make ready iterates

k-apol committed:
Refactor test to correctly retry topic creation in makeready

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java:
```diff
@@ -480,6 +482,11 @@ public Set<String> makeReady(final Map<String, InternalTopicConfig> topics) {
                     .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
             final Set<NewTopic> topicsToCreate = computeTopicsToCreate(notReadyTopicsMap, tempUnknownTopics);
+
+            topicsNotReady.retainAll(tempUnknownTopics);
```
Why do we remove unknown topics? If a topic is unknown it is by definition not ready, and we should keep it in `topicsNotReady`.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java:
```diff
@@ -480,6 +482,11 @@ public Set<String> makeReady(final Map<String, InternalTopicConfig> topics) {
+
+            topicsNotReady.retainAll(tempUnknownTopics);
+            topicsNotReady.retainAll(notReadyTopicsMap.keySet());
+            topicsToCreate.forEach(topic -> topicsNotReady.add(topic.name()));
```
This seems to be unnecessarily complex? We first remove topics, and in the third line add back topics... So it seem we are removing topic we should not have remove, just to add them back?

In the end, we should only remove the topics from `topicNotReady` which do already exist.

The issue seems to be, that we want to retain both `tempUnknownTopics` and `notReadyTopicsMap.keySet()`, but calling `retainAll` twice does not do this, but we would need a call `topicsNotReady.retainAll(union(tempUnknownTopics, notReadyTopicsMap.keySet()))`?

Or we rewrite this using `removeIf` instead of the three lines you added (I believe [but not 100% sure] this line would do the trick?
```
topicsNotReady.removeIf(t -> !topicsToCreate.contains(t) && !tempUnknownTopics.contains(t));
```

Or maybe my original comment could help (or would this not work?) https://github.com/apache/kafka/pull/20326#discussion_r2512435233

I think, if we modify `topicsNotReady` there (ie, `topicMap` which is the corresponding parameter name inside the `computeTopicsToCreate` for which we pass in `topicsNotReady`), it should fix it, allowing us to avoid the code complexity here?

k-apol replied:
fixed with recent commit, thanks for deep-dive on this

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java:
```diff
@@ -421,9 +426,11 @@ public Map<String, List<TopicPartitionInfo>> getTopicPartitionInfo(final Set<Str
+            final Map<String, List<TopicPartitionInfo>> existed = getTopicPartitionInfo(topicsToDescribe, tempUnknownTopics);
             topicPartitionInfo.putAll(existed);
             topicsToDescribe.removeAll(topicPartitionInfo.keySet());
+            topicsToDescribe.addAll(tempUnknownTopics); // keep retrying unknown ones
```
It's not clear to me, why we need this change and keep retrying?

k-apol replied:
My thinking here was that a transient error could potentially mask topics that would not be ready in this scenario.. looking at it now, it is excessive, as a network call to the broker would have succeeded at this point. I will remove it as I am not confident in this change myself now

k-apol committed:
Fix issues with recognition of which topic to keep and remove

k-apol committed:
remove retrying of unknown topics - not needed

mjsax commented:
Thanks for this refactoring! Merged to `trunk`. Great improvement over the previous state of this code!

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java:
```diff
@@ -461,121 +467,159 @@ public Set<String> makeReady(final Map<String, InternalTopicConfig> topics) {
             if (!topicsNotReady.isEmpty()) {
-                final Set<NewTopic> newTopics = new HashSet<>();
+                maybeThrowTimeout(new TimeoutContext(
+                    Collections.singleton("makeReadyCheck"), // dummy collection just to trigger if `topicsNotReady` is non-empty
```
@k-apol -- why do we pass in this dummy instead of `topicsNotReady` ?

k-apol replied:
Thanks for catching this, originally I was going to use this approach when it WAS empty but still timing out, and failed to update this error message as the check for that shifted around. I can make an update for this

mjsax replied:
Thanks. Did a small follow up PR for it: https://github.com/apache/kafka/pull/21165

k-apol commented:
It's just to satisfy the error object contract

On Tue, Dec 16, 2025, 11:45 AM Matthias J. Sax ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In
> streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java
> <https://github.com/apache/kafka/pull/20326#discussion_r2624191689>:
>
> >              if (!topicsNotReady.isEmpty()) {
> -                final Set<NewTopic> newTopics = new HashSet<>();
> +                maybeThrowTimeout(new TimeoutContext(
> +                    Collections.singleton("makeReadyCheck"), // dummy collection just to trigger if `topicsNotReady` is non-empty
>
> @k-apol <https://github.com/k-apol> -- why do we pass in this dummy
> instead of topicsNotReady ?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/kafka/pull/20326#pullrequestreview-3584266771>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BROWOLHK5HPY54DG6CZ6CYL4CBAKNAVCNFSM6AAAAACDNWOXOSVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZTKOBUGI3DMNZXGE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>
